### PR TITLE
Allow user-defined block explorers

### LINKED ISSUE #4371: Support setting one or more user-defined block explorers
Currently Bisq supports a fixed set of options for which block explorer to use when following links in the application to Bitcoin transactions and addresses. These options are hard-coded at https://github.com/bisq-network/bisq/blob/release/v1.3.6/core/src/main/java/bisq/core/user/Preferences.java#L80-L99 and show up in the application here:

image

Use of any centralized block explorer is a known privacy risk, because the service may be logging queries, including the user's IP address and other unique information and linking them to the Bitcoin addresses and/or transactions being queried. This risk can be mitigated by using a Tor-based block explorer service, but such use is still not ideal. See [these](https://en.bitcoin.it/Privacy#Blockchain_explorer_websites) [two](https://en.bitcoin.it/Privacy#Bad_privacy_example_-_Using_a_blockchain_explorer) sections of @chris-belcher's Privacy article in the Bitcoin Wiki's for more details.

The ideal approach is to use a private block explorer, i.e. one running on one's own local network that queries one's own Bitcoin Core node. Doing so eliminates the risk of being tracked by other parties as no other parties are involved.

It is easy enough to add one's own block explorer by modifying `Preferences.java` (see link above), but ideally one would be able to specify one or more custom block explorer URLs in the `Settings->Preferences` screen above. A precedent for this kind of configuration functionality can be found in the `Settings->Network Info` screen:

image

Note how, if the user selects the 'Use custom Bitcoin Core nodes' radio button, they can then provide one or more custom node addresses in the input field below. The interaction would need to be different in the case of specifying custom block explorers, because each block explorer requires two URL templates, one for transaction lookups and one for address lookups, but this existing, somewhat similar configuration functionality may be useful as a starting point when considering how to implement this feature.

While it might be easiest to support adding just one custom block explorer, it would be ideal to allow specifying multiple of them, because in the case of an explorer running on one's local network, one might wish to access it via clearnet http when at home and via Tor when remote. This would require specifying two sets of URLs. And of course the user might want to specify any number of additional custom explorer services.

### PR DESCRIPTION
**Pull request by ghost**
In Settings/General Preferences the Bitcoin Explorer and Bisq Explorer may now be user defined in addition to the hard coded block explorer choices.

This implementation takes advantage of the existing Protobuf implementation that stores a BlockExplorer Name/UrlTx/UrlAddress for BTC and for BSQ.

A new popup dialog is implemented which allows editing of the explorer details (name, txUrl, and addressUrl).  Any of the system-defined explorer details may be selected into the edit fields thus allowing them to be used as templates, customized or an entirely new explorer to be used.

If the user switches choice from a customized back to a pre-defined explorer then the customized settings are gone; this is because the Protobuf storage only allows saving one selection.

Fixes #4371
Closes #4634

image

image

### CODE REVIEW

**Comment by ghost** [comment]
Fixes #4634

**Comment by cd2357** [comment]
ACK, tested and works as expected.

File: desktop/src/main/java/bisq/desktop/main/settings/preferences/PreferencesView.java [hunk 1]
```diff
@@ -937,6 +906,34 @@ private void activateDisplayPreferences() {
 
         resetDontShowAgainButton.setOnAction(e -> preferences.resetDontShowAgain());
 
+        editCustomBtcExplorer.setOnAction(e -> {
+            EditCustomExplorerWindow urlWindow = new EditCustomExplorerWindow("BTC",
+                    preferences.getBlockChainExplorer(), preferences.getBlockChainExplorers());
+            urlWindow
+                    .actionButtonText("Save")
```
Comment (chimp1984) [inline desktop/src/main/java/bisq/desktop/main/settings/preferences/PreferencesView.java]: I think you forgot here to use a resource string. I guess its: `Res.get("shared.save")`

File: desktop/src/main/java/bisq/desktop/main/settings/preferences/PreferencesView.java [hunk 2]
```diff
@@ -937,6 +906,34 @@ private void activateDisplayPreferences() {
...
+                    .closeButtonText(Res.get("shared.cancel"))
+                    .onClose(urlWindow::hide)
+                    .show();
+        });
+
+        editCustomBsqExplorer.setOnAction(e -> {
+            EditCustomExplorerWindow urlWindow = new EditCustomExplorerWindow("BSQ",
+                    preferences.getBsqBlockChainExplorer(), preferences.getBsqBlockChainExplorers());
+            urlWindow
+                    .actionButtonText("Save")
```
Comment (chimp1984) [inline desktop/src/main/java/bisq/desktop/main/settings/preferences/PreferencesView.java]: I think you forgot here to use a resource string. I guess its: Res.get("shared.save")

**Review by chimp1984 — DISMISSED** [review]
utACK

Beside 2 nits... See comments

**Review by chimp1984 — DISMISSED** [review]
utACK

File: desktop/src/main/java/bisq/desktop/main/overlays/windows/EditCustomExplorerWindow.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+        Button button = new Button(">>");
+        button.setOnAction(e -> {
+            BlockChainExplorer blockChainExplorer = listView.getSelectionModel().getSelectedItem();
+            if (blockChainExplorer != null) {
+                nameInputTextField.setText(blockChainExplorer.name);
+                txUrlInputTextField.setText(blockChainExplorer.txUrl);
+                addressUrlInputTextField.setText(blockChainExplorer.addressUrl);
+            }
+        });
```
Comment (ripcurlx) [inline desktop/src/main/java/bisq/desktop/main/overlays/windows/EditCustomExplorerWindow.java]: Please use our AutoTooltipButton or JFXButton to match the design of all other buttons.

**Review by ripcurlx — CHANGES_REQUESTED** [review]
NACK

Code looks fine, but please change the style of the `>>` button to match our general button style.

File: desktop/src/main/java/bisq/desktop/main/overlays/windows/EditCustomExplorerWindow.java (hunk 3 above)
Comment (ghost) [reply desktop/src/main/java/bisq/desktop/main/overlays/windows/EditCustomExplorerWindow.java]: :+1: Done!

**Comment by sqrrm** [comment]
@ripcurlx Are you ok with this PR now? Will leave merge for you

**Review by ripcurlx — APPROVED** [review]

**Comment by cbeams** [comment]
Great to see this functionality, thanks @jmacxx!

One note for potential future contributions: it would be great to be able to have more than one custom explorer. This is from my original description of the feature at https://github.com/bisq-network/bisq/issues/4371#issue-664425011:

> While it might be easiest to support adding just one custom block explorer, it would be ideal to allow specifying multiple of them, because in the case of an explorer running on one's local network, one might wish to access it via clearnet http when at home and via Tor when remote. This would require specifying two sets of URLs. And of course the user might want to specify any number of additional custom explorer services.

**Pull request merged**